File: demos/demo_filter_image_weighted.m
%%% Example script showing how to perform a weighted 2D Total-Variation filtering with proxTV

clear all
close all

% Prepare plot
figure()
colormap gray

% Load image
X = rgb2gray(imread('colors.png'));
X=double(X)/255;

% Introduce noise
noiseLevel = 0.01;
N = double(imnoise(X,'speckle'));

% Gradient in columns
W1=0.01*cumsum(ones(size(X,1)-1, size(X,2)),2);
W2=0.01*ones(size(X,1), size(X,2)-1);
disp('Solving 2D weighted TV...');
tic;
FW = TV(N,{W1,W2});
toc;
subplot(3,4,1); imagesc(W1); title('Weights along columns');
subplot(3,4,5); imagesc(W2); title('Weights along rows');
subplot(3,4,9); imagesc(FW); title('Filter result');

% Gradient in rows
W1=0.01*ones(size(X,1)-1, size(X,2));
W2=0.01*cumsum(ones(size(X,1), size(X,2)-1),1);
disp('Solving 2D weighted TV...');
tic;
FW = TV(N,{W1,W2});
toc;
subplot(3,4,2); imagesc(W1); title('Weights along columns');
subplot(3,4,6); imagesc(W2); title('Weights along rows');
subplot(3,4,10); imagesc(FW); title('Filter result');

% Gradient in columnas and rows
W1=0.004*cumsum(ones(size(X,1)-1, size(X,2)),2);
W2=0.004*cumsum(ones(size(X,1), size(X,2)-1),1);
disp('Solving 2D weighted TV...');
tic;
FW = TV(N,{W1,W2});
toc;
subplot(3,4,3); imagesc(W1); title('Weights along columns');
subplot(3,4,7); imagesc(W2); title('Weights along rows');
subplot(3,4,11); imagesc(FW); title('Filter result')

% Grid regions
W1=max(0,cos(0.05*cumsum(ones(size(X,1)-1, size(X,2)),2)));
W2=max(0,sin(0.05*cumsum(ones(size(X,1), size(X,2)-1),1)));
disp('Solving 2D weighted TV...');
tic;
FW = TV(N,{W1,W2});
toc;
subplot(3,4,4); imagesc(W1); title('Weights along columns');
subplot(3,4,8); imagesc(W2); title('Weights along rows');
subplot(3,4,12); imagesc(FW); title('Filter result')

File: prox_tv/demos/demo_filter_image_weighted.py
### Example script showing how to perform a weighted 2D Total-Variation filtering with proxTV
import prox_tv as ptv
import numpy as np
from pylab import *
import matplotlib.pyplot as plt
import time
import skimage as ski
from skimage import data, io, filters, color, util

# Load image
X = io.imread('colors.png')
X = ski.img_as_float(X)
X = color.rgb2gray(X)

# Introduce noise
noiseLevel = 0.01
N = util.random_noise(X, mode='speckle', var=noiseLevel)

# Gradient in columns
W1 = 0.01 * np.cumsum(np.ones((X.shape[0]-1, X.shape[1])), 1)
W2 = 0.01 * np.ones((X.shape[0], X.shape[1]-1))
print('Solving 2D weighted TV...' )
start = time.time()
FW = ptv.tv1w_2d(N, W1, W2)
end = time.time()
print('Elapsed time ' + str(end-start))

plt.subplot(3, 4, 1); io.imshow(W1); plt.title('Weights along columns');
plt.subplot(3, 4, 5); io.imshow(W2); plt.title('Weights along rows');
plt.subplot(3, 4, 9); io.imshow(FW); plt.title('Filter result');

# Gradient in rows
W1 = 0.01 * np.ones((X.shape[0]-1, X.shape[1]))
W2 = 0.01 * np.cumsum(np.ones((X.shape[0], X.shape[1]-1)), 0)
print('Solving 2D weighted TV...')
start = time.time()
FW = ptv.tv1w_2d(N, W1, W2)
end = time.time()
print('Elapsed time ' + str(end-start))

plt.subplot(3, 4, 2); io.imshow(W1); plt.title('Weights along columns');
plt.subplot(3, 4, 6); io.imshow(W2); plt.title('Weights along rows');
plt.subplot(3, 4, 10); io.imshow(FW); plt.title('Filter result');

# Gradient in columns and rows
W1 = 0.004 * np.cumsum(np.ones((X.shape[0]-1, X.shape[1])), 1)
W2 = 0.004 * np.cumsum(np.ones((X.shape[0], X.shape[1]-1)), 0)
print('Solving 2D weighted TV...' )
start = time.time()
FW = ptv.tv1w_2d(N, W1, W2)
end = time.time()
print('Elapsed time ' + str(end-start))

plt.subplot(3, 4, 3); io.imshow(W1); plt.title('Weights along columns');
plt.subplot(3, 4, 7); io.imshow(W2); plt.title('Weights along rows');
plt.subplot(3, 4, 11); io.imshow(FW); plt.title('Filter result');

# Grid regions
W1=cos(0.05*np.cumsum(np.ones((X.shape[0]-1, X.shape[1])), 1))
W1[W1 < 0] = 0
W2=sin(0.05*np.cumsum(np.ones((X.shape[0], X.shape[1]-1)), 0))
W2[W2 < 0] = 0;
print('Solving 2D weighted TV...')
start = time.time()
FW = ptv.tv1w_2d(N, W1, W2)
end = time.time()
print('Elapsed time ' + str(end-start))

plt.subplot(3, 4, 4); io.imshow(W1); plt.title('Weights along columns');
plt.subplot(3, 4, 8); io.imshow(W2); plt.title('Weights along rows');
plt.subplot(3, 4, 12); io.imshow(FW); plt.title('Filter result');

plt.show()

